src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -722,12 +788,28 @@ private Task DecodeHeadersAsync<TContext>(IHttpApplication<TContext> application
 
                 if (endHeaders)
                 {
-                    StartStream(application);
-                    ResetRequestHeaderParsingState();
+                    lock (_connectionStateLock)
+                    {
+                        if (_connectionState != ConnectionState.Closed)
+                        {
+                            StartStream(application);
+                            _highestOpenedStreamId = _currentHeadersStream.StreamId;
+                        }
+
+                        ResetRequestHeaderParsingState();
+                    }
                 }
             }
             catch (Http2StreamErrorException)
             {
+                lock (_connectionStateLock)
+                {
+                    if (_connectionState != ConnectionState.Closed)
```

Comment:
Why do you acquire the lock twice in this method? Wouldn't it read better to acquire a lock around the entire method body and go back to setting _highestOpenedStreamId in ResetRequestHeaderParsingState()? Is it even important to call ResetRequestHeaderParsingState() if the connection is already in the closed state?

Reply:
Trying to keep the locked operations as few as possible. I don't see why resetting the state would be necessary in a closed state but it seems like the safe thing to do.

Reply:
I don't mind resetting the state in the closed state so much. Acquiring the lock twice in the same method seems unnecessary in this case though. Just lock around the whole thing.